File: Dynamic_Programming/JavaScript/knapsack_problem.js
/**
 * @file knapsack_problem.js
 * @description Implementation of Knapsack Problem algorithm in JavaScript
 * @author Algorithm Collection
 * @version 1.0
 */

/**
 * @function knapsack01Recursive
 * @description Recursive implementation of 0/1 Knapsack
 * @param {Array<number>} weights - Array of item weights
 * @param {Array<number>} values - Array of item values
 * @param {number} capacity - Knapsack capacity
 * @returns {number} - Maximum value that can be obtained
 */
function knapsack01Recursive(weights, values, capacity) {
    return knapsack01RecursiveHelper(weights, values, capacity, weights.length);
}

/**
 * @function knapsack01RecursiveHelper
 * @description Helper function for recursive 0/1 Knapsack
 * @param {Array<number>} weights - Array of item weights
 * @param {Array<number>} values - Array of item values
 * @param {number} capacity - Remaining capacity
 * @param {number} n - Number of items to consider
 * @returns {number} - Maximum value that can be obtained
 */
function knapsack01RecursiveHelper(weights, values, capacity, n) {
    if (n === 0 || capacity === 0) {
        return 0;
    }
    
    if (weights[n - 1] > capacity) {
        return knapsack01RecursiveHelper(weights, values, capacity, n - 1);
    }
    
    return Math.max(
        values[n - 1] + knapsack01RecursiveHelper(weights, values, capacity - weights[n - 1], n - 1),
        knapsack01RecursiveHelper(weights, values, capacity, n - 1)
    );
}

/**
 * @function knapsack01DP
 * @description Dynamic programming implementation of 0/1 Knapsack
 * @param {Array<number>} weights - Array of item weights
 * @param {Array<number>} values - Array of item values
 * @param {number} capacity - Knapsack capacity
 * @returns {number} - Maximum value that can be obtained
 */
function knapsack01DP(weights, values, capacity) {
    const n = weights.length;
    const dp = Array(n + 1).fill().map(() => Array(capacity + 1).fill(0));
    
    for (let i = 1; i <= n; i++) {
        for (let w = 1; w <= capacity; w++) {
            if (weights[i - 1] <= w) {
                dp[i][w] = Math.max(
                    values[i - 1] + dp[i - 1][w - weights[i - 1]],
                    dp[i - 1][w]
                );
            } else {
                dp[i][w] = dp[i - 1][w];
            }
        }
    }
    
    return dp[n][capacity];
}

/**
 * @function knapsack01Optimized
 * @description Space-optimized implementation of 0/1 Knapsack
 * @param {Array<number>} weights - Array of item weights
 * @param {Array<number>} values - Array of item values
 * @param {number} capacity - Knapsack capacity
 * @returns {number} - Maximum value that can be obtained
 */
function knapsack01Optimized(weights, values, capacity) {
    const dp = new Array(capacity + 1).fill(0);
    
    for (let i = 0; i < weights.length; i++) {
        for (let w = capacity; w >= weights[i]; w--) {
            dp[w] = Math.max(dp[w], dp[w - weights[i]] + values[i]);
        }
    }
    
    return dp[capacity];
}

/**
 * @function knapsack01WithItems
 * @description 0/1 Knapsack that returns selected items
 * @param {Array<number>} weights - Array of item weights
 * @param {Array<number>} values - Array of item values
 * @param {number} capacity - Knapsack capacity
 * @returns {Array<number>} - List of selected item indices
 */
function knapsack01WithItems(weights, values, capacity) {
    const n = weights.length;
    const dp = Array(n + 1).fill().map(() => Array(capacity + 1).fill(0));
    
    // Fill DP table
    for (let i = 1; i <= n; i++) {
        for (let w = 1; w <= capacity; w++) {
            if (weights[i - 1] <= w) {
                dp[i][w] = Math.max(
                    values[i - 1] + dp[i - 1][w - weights[i - 1]],
                    dp[i - 1][w]
                );
            } else {
                dp[i][w] = dp[i - 1][w];
            }
        }
    }
    
    // Backtrack to find selected items
    const selectedItems = [];
    let i = n, w = capacity;
    
    while (i > 0 && w > 0) {
        if (dp[i][w] !== dp[i - 1][w]) {
            selectedItems.push(i - 1);
            w -= weights[i - 1];
        }
        i--;
    }
    
    return selectedItems.reverse();
}

/**
 * @function knapsackFractional
 * @description Fractional knapsack using greedy approach
 * @param {Array<number>} weights - Array of item weights
 * @param {Array<number>} values - Array of item values
 * @param {number} capacity - Knapsack capacity
 * @returns {number} - Maximum value that can be obtained
 */
function knapsackFractional(weights, values, capacity) {
    const n = weights.length;
    const items = [];
    
    for (let i = 0; i < n; i++) {
        items.push({
            index: i,
            weight: weights[i],
            value: values[i],
            valuePerWeight: values[i] / weights[i]
        });
    }
    
    // Sort by value per unit weight
    items.sort((a, b) => b.valuePerWeight - a.valuePerWeight);
    
    let totalValue = 0;
    let remainingCapacity = capacity;
    
    for (const item of items) {
        if (remainingCapacity >= item.weight) {
            totalValue += item.value;
            remainingCapacity -= item.weight;
        } else {
            totalValue += item.valuePerWeight * remainingCapacity;
            break;
        }
    }
    
    return totalValue;
}

/**
 * @function knapsackUnbounded
 * @description Unbounded knapsack problem
 * @param {Array<number>} weights - Array of item weights
 * @param {Array<number>} values - Array of item values
 * @param {number} capacity - Knapsack capacity
 * @returns {number} - Maximum value that can be obtained
 */
function knapsackUnbounded(weights, values, capacity) {
    const dp = new Array(capacity + 1).fill(0);
    
    for (let w = 1; w <= capacity; w++) {
        for (let i = 0; i < weights.length; i++) {
            if (weights[i] <= w) {
                dp[w] = Math.max(dp[w], dp[w - weights[i]] + values[i]);
            }
        }
    }
    
    return dp[capacity];
}

/**
 * @function knapsackMultiple
 * @description Multiple knapsack problem
 * @param {Array<number>} weights - Array of item weights
 * @param {Array<number>} values - Array of item values
 * @param {Array<number>} capacities - Array of knapsack capacities
 * @returns {number} - Maximum value that can be obtained
 */
function knapsackMultiple(weights, values, capacities) {
    const n = weights.length;
    const k = capacities.length;
    const dp = Array(n + 1).fill().map(() => 
        Array(k + 1).fill().map(() => new Array(Math.max(...capacities) + 1).fill(0))
    );
    
    for (let i = 1; i <= n; i++) {
        for (let j = 1; j <= k; j++) {
            for (let w = 1; w <= capacities[j - 1]; w++) {
                if (weights[i - 1] <= w) {
                    dp[i][j][w] = Math.max(
                        values[i - 1] + dp[i - 1][j][w - weights[i - 1]],
                        dp[i - 1][j][w]
                    );
                } else {
                    dp[i][j][w] = dp[i - 1][j][w];
                }
            }
        }
    }
    
    return dp[n][k][capacities[k - 1]];
}

/**
 * @function createSampleData
 * @description Creates sample data for testing
 * @returns {Object} - Sample data
 */
function createSampleData() {
    return {
        weights: [10, 20, 30, 40, 50],
        values: [60, 100, 120, 160, 200],
        capacity: 100
    };
}

/**
 * @function testKnapsack
 * @description Test function to demonstrate knapsack algorithms
 * @returns {void}
 */
function testKnapsack() {
    console.log("=== Knapsack Problem Test ===");
    
    const data = createSampleData();
    const { weights, values, capacity } = data;
    
    console.log("Sample data:");
    console.log("Weights:", weights);
    console.log("Values:", values);
    console.log("Capacity:", capacity);
    console.log();
    
    // Test 0/1 Knapsack - Recursive
    console.log("0/1 Knapsack - Recursive:");
    const startTime = performance.now();
    const recursiveResult = knapsack01Recursive(weights, values, capacity);
    const endTime = performance.now();
    console.log(`Result: ${recursiveResult} (${(endTime - startTime).toFixed(2)} ms)`);
    console.log();
    
    // Test 0/1 Knapsack - DP
    console.log("0/1 Knapsack - DP:");
    const startTime2 = performance.now();
    const dpResult = knapsack01DP(weights, values, capacity);
    const endTime2 = performance.now();
    console.log(`Result: ${dpResult} (${(endTime2 - startTime2).toFixed(2)} ms)`);
    console.log();
    
    // Test 0/1 Knapsack - Optimized
    console.log("0/1 Knapsack - Optimized:");
    const startTime3 = performance.now();
    const optimizedResult = knapsack01Optimized(weights, values, capacity);
    const endTime3 = performance.now();
    console.log(`Result: ${optimizedResult} (${(endTime3 - startTime3).toFixed(2)} ms)`);
    console.log();
    
    // Test 0/1 Knapsack with items
    console.log("0/1 Knapsack - Selected items:");
    const selectedItems = knapsack01WithItems(weights, values, capacity);
    console.log("Selected items:", selectedItems);
    console.log();
    
    // Test Fractional Knapsack
    console.log("Fractional Knapsack:");
    const startTime4 = performance.now();
    const fractionalResult = knapsackFractional(weights, values, capacity);
    const endTime4 = performance.now();
    console.log(`Result: ${fractionalResult.toFixed(2)} (${(endTime4 - startTime4).toFixed(2)} ms)`);
    console.log();
    
    // Test Unbounded Knapsack
    console.log("Unbounded Knapsack:");
    const startTime5 = performance.now();
    const unboundedResult = knapsackUnbounded(weights, values, capacity);
    const endTime5 = performance.now();
    console.log(`Result: ${unboundedResult} (${(endTime5 - startTime5).toFixed(2)} ms)`);
    console.log();
}

/**
 * @function performanceTest
 * @description Performance test for knapsack algorithms
 * @returns {void}
 */
function performanceTest() {
    console.log("=== Performance Test ===");
    
    const sizes = [50, 100, 200, 500, 1000];
    
    for (const size of sizes) {
        const weights = new Array(size);
        const values = new Array(size);
        const capacity = size * 10;
        
        // Generate random data
        for (let i = 0; i < size; i++) {
            weights[i] = Math.floor(Math.random() * 50) + 1;
            values[i] = Math.floor(Math.random() * 100) + 1;
        }
        
        console.log(`Size: ${size}, Capacity: ${capacity}`);
        
        // Test DP
        const startTime = performance.now();
        knapsack01DP(weights, values, capacity);
        const endTime = performance.now();
        console.log(`  DP: ${(endTime - startTime).toFixed(2)} ms`);
        
        // Test Optimized
        const startTime2 = performance.now();
        knapsack01Optimized(weights, values, capacity);
        const endTime2 = performance.now();
        console.log(`  Optimized: ${(endTime2 - startTime2).toFixed(2)} ms`);
        
        console.log();
    }
}

/**
 * @function main
 * @description Main function to run the knapsack demonstration
 * @returns {void}
 */
function main() {
    console.log("Knapsack Problem Algorithm Implementation");
    console.log("========================================");
    
    // Run basic tests
    testKnapsack();
    
    // Run performance test
    performanceTest();
    
    console.log("\nKnapsack problem completed successfully!");
}

// Export functions for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        knapsack01Recursive,
        knapsack01DP,
        knapsack01Optimized,
        knapsack01WithItems,
        knapsackFractional,
        knapsackUnbounded,
        knapsackMultiple,
        createSampleData,
        testKnapsack,
        performanceTest,
        main
    };
}

// Run the main function if this file is executed directly
if (typeof require !== 'undefined' && require.main === module) {
    main();
}

File: Dynamic_Programming/Java/knapsack_problem.java

/**
 * @file knapsack_problem.java
 * @description Implementation of Knapsack Problem algorithm in Java
 * @author Algorithm Collection
 * @version 1.0
 */

import java.util.*;

/**
 * @class knapsack_problem
 * @description Class containing Knapsack Problem algorithm implementation
 */
public class knapsack_problem {

    /**
     * @function knapsack01Recursive
     * @description Recursive implementation of 0/1 Knapsack
     * @param weights  int[] - Array of item weights
     * @param values   int[] - Array of item values
     * @param capacity int - Knapsack capacity
     * @returns int - Maximum value that can be obtained
     */
    public static int knapsack01Recursive(int[] weights, int[] values, int capacity) {
        return knapsack01RecursiveHelper(weights, values, capacity, weights.length);
    }

    /**
     * @function knapsack01RecursiveHelper
     * @description Helper function for recursive 0/1 Knapsack
     * @param weights  int[] - Array of item weights
     * @param values   int[] - Array of item values
     * @param capacity int - Remaining capacity
     * @param n        int - Number of items to consider
     * @returns int - Maximum value that can be obtained
     */
    private static int knapsack01RecursiveHelper(int[] weights, int[] values, int capacity, int n) {
        if (n == 0 || capacity == 0) {
            return 0;
        }

        if (weights[n - 1] > capacity) {
            return knapsack01RecursiveHelper(weights, values, capacity, n - 1);
        }

        return Math.max(
                values[n - 1] + knapsack01RecursiveHelper(weights, values, capacity - weights[n - 1], n - 1),
                knapsack01RecursiveHelper(weights, values, capacity, n - 1));
    }

    /**
     * @function knapsack01DP
     * @description Dynamic programming implementation of 0/1 Knapsack
     * @param weights  int[] - Array of item weights
     * @param values   int[] - Array of item values
     * @param capacity int - Knapsack capacity
     * @returns int - Maximum value that can be obtained
     */
    public static int knapsack01DP(int[] weights, int[] values, int capacity) {
        int n = weights.length;
        int[][] dp = new int[n + 1][capacity + 1];

        for (int i = 1; i <= n; i++) {
            for (int w = 1; w <= capacity; w++) {
                if (weights[i - 1] <= w) {
                    dp[i][w] = Math.max(
                            values[i - 1] + dp[i - 1][w - weights[i - 1]],
                            dp[i - 1][w]);
                } else {
                    dp[i][w] = dp[i - 1][w];
                }
            }
        }

        return dp[n][capacity];
    }

    /**
     * @function knapsack01Optimized
     * @description Space-optimized implementation of 0/1 Knapsack
     * @param weights  int[] - Array of item weights
     * @param values   int[] - Array of item values
     * @param capacity int - Knapsack capacity
     * @returns int - Maximum value that can be obtained
     */
    public static int knapsack01Optimized(int[] weights, int[] values, int capacity) {
        int[] dp = new int[capacity + 1];

        for (int i = 0; i < weights.length; i++) {
            for (int w = capacity; w >= weights[i]; w--) {
                dp[w] = Math.max(dp[w], dp[w - weights[i]] + values[i]);
            }
        }

        return dp[capacity];
    }

    /**
     * @function knapsack01WithItems
     * @description 0/1 Knapsack that returns selected items
     * @param weights  int[] - Array of item weights
     * @param values   int[] - Array of item values
     * @param capacity int - Knapsack capacity
     * @returns List<Integer> - List of selected item indices
     */
    public static List<Integer> knapsack01WithItems(int[] weights, int[] values, int capacity) {
        int n = weights.length;
        int[][] dp = new int[n + 1][capacity + 1];

        // Fill DP table
        for (int i = 1; i <= n; i++) {
            for (int w = 1; w <= capacity; w++) {
                if (weights[i - 1] <= w) {
                    dp[i][w] = Math.max(
                            values[i - 1] + dp[i - 1][w - weights[i - 1]],
                            dp[i - 1][w]);
                } else {
                    dp[i][w] = dp[i - 1][w];
                }
            }
        }

        // Backtrack to find selected items
        List<Integer> selectedItems = new ArrayList<>();
        int i = n, w = capacity;

        while (i > 0 && w > 0) {
            if (dp[i][w] != dp[i - 1][w]) {
                selectedItems.add(i - 1);
                w -= weights[i - 1];
            }
            i--;
        }

        Collections.reverse(selectedItems);
        return selectedItems;
    }

    /**
     * @function knapsackFractional
     * @description Fractional knapsack using greedy approach
     * @param weights  int[] - Array of item weights
     * @param values   int[] - Array of item values
     * @param capacity int - Knapsack capacity
     * @returns double - Maximum value that can be obtained
     */
    public static double knapsackFractional(int[] weights, int[] values, int capacity) {
        int n = weights.length;
        Item[] items = new Item[n];

        for (int i = 0; i < n; i++) {
            items[i] = new Item(i, weights[i], values[i]);
        }

        // Sort by value per unit weight
        Arrays.sort(items, (a, b) -> Double.compare(b.valuePerWeight, a.valuePerWeight));

        double totalValue = 0;
        int remainingCapacity = capacity;

        for (Item item : items) {
            if (remainingCapacity >= item.weight) {
                totalValue += item.value;
                remainingCapacity -= item.weight;
            } else {
                totalValue += item.valuePerWeight * remainingCapacity;
                break;
            }
        }

        return totalValue;
    }

    /**
     * @class Item
     * @description Represents an item in the knapsack
     */
    static class Item {
        int index;
        int weight;
        int value;
        double valuePerWeight;

        Item(int index, int weight, int value) {
            this.index = index;
            this.weight = weight;
            this.value = value;
            this.valuePerWeight = (double) value / weight;
        }
    }

    /**
     * @function knapsackUnbounded
     * @description Unbounded knapsack problem
     * @param weights  int[] - Array of item weights
     * @param values   int[] - Array of item values
     * @param capacity int - Knapsack capacity
     * @returns int - Maximum value that can be obtained
     */
    public static int knapsackUnbounded(int[] weights, int[] values, int capacity) {
        int[] dp = new int[capacity + 1];

        for (int w = 1; w <= capacity; w++) {
            for (int i = 0; i < weights.length; i++) {
                if (weights[i] <= w) {
                    dp[w] = Math.max(dp[w], dp[w - weights[i]] + values[i]);
                }
            }
        }

        return dp[capacity];
    }

    /**
     * @function knapsackMultiple
     * @description Multiple knapsack problem
     * @param weights    int[] - Array of item weights
     * @param values     int[] - Array of item values
     * @param capacities int[] - Array of knapsack capacities
     * @returns int - Maximum value that can be obtained
     */
    public static int knapsackMultiple(int[] weights, int[] values, int[] capacities) {
        int n = weights.length;
        int k = capacities.length;
        int[][][] dp = new int[n + 1][k + 1][];

        // Initialize DP table
        for (int i = 0; i <= n; i++) {
            for (int j = 0; j <= k; j++) {
                dp[i][j] = new int[capacities[j] + 1];
            }
        }

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= k; j++) {
                for (int w = 1; w <= capacities[j - 1]; w++) {
                    if (weights[i - 1] <= w) {
                        dp[i][j][w] = Math.max(
                                values[i - 1] + dp[i - 1][j][w - weights[i - 1]],
                                dp[i - 1][j][w]);
                    } else {
                        dp[i][j][w] = dp[i - 1][j][w];
                    }
                }
            }
        }

        return dp[n][k][capacities[k - 1]];
    }

    /**
     * @function createSampleData
     * @description Creates sample data for testing
     * @returns Map<String, Object> - Sample data
     */
    public static Map<String, Object> createSampleData() {
        Map<String, Object> data = new HashMap<>();

        int[] weights = { 10, 20, 30, 40, 50 };
        int[] values = { 60, 100, 120, 160, 200 };
        int capacity = 100;

        data.put("weights", weights);
        data.put("values", values);
        data.put("capacity", capacity);

        return data;
    }

    /**
     * @function testKnapsack
     * @description Test function to demonstrate knapsack algorithms
     * @returns void
     */
    public static void testKnapsack() {
        System.out.println("=== Knapsack Problem Test ===");

        Map<String, Object> data = createSampleData();
        int[] weights = (int[]) data.get("weights");
        int[] values = (int[]) data.get("values");
        int capacity = (int) data.get("capacity");

        System.out.println("Sample data:");
        System.out.println("Weights: " + Arrays.toString(weights));
        System.out.println("Values: " + Arrays.toString(values));
        System.out.println("Capacity: " + capacity);
        System.out.println();

        // Test 0/1 Knapsack - Recursive
        System.out.println("0/1 Knapsack - Recursive:");
        long startTime = System.nanoTime();
        int recursiveResult = knapsack01Recursive(weights, values, capacity);
        long endTime = System.nanoTime();
        System.out.printf("Result: %d (%.2f ms)%n", recursiveResult, (endTime - startTime) / 1_000_000.0);
        System.out.println();

        // Test 0/1 Knapsack - DP
        System.out.println("0/1 Knapsack - DP:");
        startTime = System.nanoTime();
        int dpResult = knapsack01DP(weights, values, capacity);
        endTime = System.nanoTime();
        System.out.printf("Result: %d (%.2f ms)%n", dpResult, (endTime - startTime) / 1_000_000.0);
        System.out.println();

        // Test 0/1 Knapsack - Optimized
        System.out.println("0/1 Knapsack - Optimized:");
        startTime = System.nanoTime();
        int optimizedResult = knapsack01Optimized(weights, values, capacity);
        endTime = System.nanoTime();
        System.out.printf("Result: %d (%.2f ms)%n", optimizedResult, (endTime - startTime) / 1_000_000.0);
        System.out.println();

        // Test 0/1 Knapsack with items
        System.out.println("0/1 Knapsack - Selected items:");
        List<Integer> selectedItems = knapsack01WithItems(weights, values, capacity);
        System.out.println("Selected items: " + selectedItems);
        System.out.println();

        // Test Fractional Knapsack
        System.out.println("Fractional Knapsack:");
        startTime = System.nanoTime();
        double fractionalResult = knapsackFractional(weights, values, capacity);
        endTime = System.nanoTime();
        System.out.printf("Result: %.2f (%.2f ms)%n", fractionalResult, (endTime - startTime) / 1_000_000.0);
        System.out.println();

        // Test Unbounded Knapsack
        System.out.println("Unbounded Knapsack:");
        startTime = System.nanoTime();
        int unboundedResult = knapsackUnbounded(weights, values, capacity);
        endTime = System.nanoTime();
        System.out.printf("Result: %d (%.2f ms)%n", unboundedResult, (endTime - startTime) / 1_000_000.0);
        System.out.println();
    }

    /**
     * @function performanceTest
     * @description Performance test for knapsack algorithms
     * @returns void
     */
    public static void performanceTest() {
        System.out.println("=== Performance Test ===");

        int[] sizes = { 50, 100, 200, 500, 1000 };

        for (int size : sizes) {
            int[] weights = new int[size];
            int[] values = new int[size];
            int capacity = size * 10;

            // Generate random data
            Random random = new Random();
            for (int i = 0; i < size; i++) {
                weights[i] = random.nextInt(50) + 1;
                values[i] = random.nextInt(100) + 1;
            }

            System.out.println("Size: " + size + ", Capacity: " + capacity);

            // Test DP
            long startTime = System.nanoTime();
            knapsack01DP(weights, values, capacity);
            long endTime = System.nanoTime();
            System.out.printf("  DP: %.2f ms%n", (endTime - startTime) / 1_000_000.0);

            // Test Optimized
            startTime = System.nanoTime();
            knapsack01Optimized(weights, values, capacity);
            endTime = System.nanoTime();
            System.out.printf("  Optimized: %.2f ms%n", (endTime - startTime) / 1_000_000.0);

            System.out.println();
        }
    }

    /**
     * @function main
     * @description Main method to run the knapsack demonstration
     * @param args String[] - Command line arguments
     * @returns void
     */
    public static void main(String[] args) {
        System.out.println("Knapsack Problem Algorithm Implementation");
        System.out.println("========================================");

        // Run basic tests
        testKnapsack();

        // Run performance test
        performanceTest();

        System.out.println("\nKnapsack problem completed successfully!");
    }
}
